File: ProjectSourceCode/src/wordpar_ai/ai_service.py
# ai_service.py
# Flask microservice with improved semantic similarity

import os
import re
from flask import Flask, request, jsonify
from flask_cors import CORS
import numpy as np
from openai import OpenAI
from sentence_transformers import SentenceTransformer, util
from dotenv import load_dotenv

# -------------------------
# Load environment
# -------------------------
load_dotenv()
OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
OPENAI_MODEL = os.getenv("OPENAI_MODEL", "gpt-4o-mini")

# -------------------------
# Initialize OpenAI + better fallback model
# -------------------------
client = OpenAI(api_key=OPENAI_API_KEY) if OPENAI_API_KEY else None

# Using a better model optimized for semantic similarity
print("Loading sentence-transformers similarity model (paraphrase-MiniLM-L6-v2)...")
local_model = SentenceTransformer("paraphrase-MiniLM-L6-v2")
print("Model loaded and ready.")

# Flask setup
app = Flask(__name__)
CORS(app)

# -------------------------
# Helpers
# -------------------------
def sanitize_one_word(text: str) -> str:
    """Return only the first token of a text, letters only, trimmed."""
    if not text:
        return ""
    cleaned = re.sub(r"[^A-Za-z\- ]+", "", text).strip()
    token = cleaned.split()[0] if cleaned.split() else ""
    return token

def cosine_sim(v1, v2):
    return float(np.dot(v1, v2) / (np.linalg.norm(v1) * np.linalg.norm(v2)))

def scale_similarity(raw_score):
    """
    Custom scaling for gameplay:
    - < 0.3: Keep low (bad guesses stay bad)
    - 0.4-0.5: +0.10-0.15 boost (decent guesses get small help)
    - 0.6-0.7: +0.20-0.30 boost (good guesses get rewarded!)
    - > 0.7: Even bigger boost for great guesses
    """
    if raw_score < 0.3:
        # Bad guesses: no boost, maybe slight penalty
        return raw_score * 0.9
    
    elif 0.3 <= raw_score < 0.4:
        # Mediocre: tiny boost
        return raw_score + 0.05
    
    elif 0.4 <= raw_score < 0.5:
        # Decent: 0.10-0.15 boost
        boost = 0.10 + (raw_score - 0.4) * 0.5  # Linear from 0.10 to 0.15
        return raw_score + boost
    
    elif 0.5 <= raw_score < 0.6:
        # Good: 0.15-0.20 boost
        boost = 0.15 + (raw_score - 0.5) * 0.5
        return raw_score + boost
    
    elif 0.6 <= raw_score < 0.7:
        # Very good: 0.20-0.30 boost (your request!)
        boost = 0.20 + (raw_score - 0.6) * 1.0  # Linear from 0.20 to 0.30
        return raw_score + boost
    
    else:  # >= 0.7
        # Excellent: 0.30+ boost
        boost = 0.30 + (raw_score - 0.7) * 0.5
        return min(raw_score + boost, 1.0)  # Cap at 1.0

# -------------------------
# Endpoint: Health
# -------------------------
@app.route("/api/health", methods=["GET"])
def health():
    return jsonify({"status": "ok"})

# -------------------------
# Endpoint: Generate Word
# -------------------------
@app.route("/api/generate_word", methods=["GET"])
def generate_word():
    """
    GET /api/generate_word?category=music
    Response: {"category": "music", "word": "Beatles", "source":"openai"}
    """
    category = (request.args.get("category") or "").strip().lower()
    if not category:
        return jsonify({"error": "category query param required"}), 400

    prompt = (
        f"Provide exactly one single English word related to the category '{category}'.\n"
        "- The word should be 4 to 9 letters long if possible, and something people might guess in a word association game.\n"
        "- Return ONLY the word, without explanation, punctuation, or quotes.\n"
    )

    if client:
        try:
            resp = client.chat.completions.create(
                model=OPENAI_MODEL,
                messages=[{"role": "user", "content": prompt}],
                temperature=0.6,
                max_tokens=10,
            )
            text = resp.choices[0].message.content
            word = sanitize_one_word(text)
            if not word:
                raise ValueError("OpenAI returned empty or invalid word")
            return jsonify({"category": category, "word": word, "source": "openai"})
        except Exception as e:
            print("OpenAI generation error:", e)
            # Fallback to simple word list
            fallback_words = {
                "music": ["guitar", "piano", "drums", "violin", "melody"],
                "sports": ["soccer", "tennis", "basketball", "swimming"],
                "food": ["pizza", "pasta", "burger", "salad"],
                "animals": ["elephant", "dolphin", "tiger", "penguin"],
            }
            word = fallback_words.get(category, ["word"])[0]
            return jsonify({
                "category": category, 
                "word": word, 
                "source": "fallback",
                "note": "OpenAI unavailable - using fallback"
            })

    return jsonify({
        "error": "OpenAI API key not set on server.",
        "how_to_fix": "Set OPENAI_API_KEY environment variable and restart.",
    }), 501

# -------------------------
# Endpoint: Similarity
# -------------------------
@app.route("/api/similarity", methods=["POST"])
def similarity():
    """
    POST /api/similarity
    JSON: {"word1": "guess", "word2": "secret"}
    Response: {"similarity": 0.742, "raw_similarity": 0.591}
    """
    data = request.get_json() or {}
    w1 = sanitize_one_word(data.get("word1", "")).lower()
    w2 = sanitize_one_word(data.get("word2", "")).lower()

    if not w1 or not w2:
        return jsonify({"error": "Both 'word1' and 'word2' required"}), 400

    try:
        # Use local model (it's better for semantic similarity than OpenAI embeddings)
        emb1 = local_model.encode(w1, convert_to_tensor=True)
        emb2 = local_model.encode(w2, convert_to_tensor=True)
        raw_score = util.cos_sim(emb1, emb2).item()
        
        # Scale for better UX
        scaled_score = scale_similarity(raw_score)
        
        return jsonify({
            "word1": w1,
            "word2": w2,
            "similarity": round(float(scaled_score), 3),
            "raw_similarity": round(float(raw_score), 3),
            "source": "local-paraphrase-model"
        })

    except Exception as e:
        print("Similarity error:", e)
        return jsonify({"error": "Similarity computation failed", "details": str(e)}), 500

# -------------------------
# Run app
# -------------------------
if __name__ == "__main__":
    port = int(os.getenv("PORT", 5000))
    print(f"Starting AI service on http://127.0.0.1:{port}")
    app.run(host="0.0.0.0", port=port, debug=True)

File: ProjectSourceCode/src/aiService.js
// src/aiService.js

import fetch from "node-fetch";

const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
const OPENAI_API_URL = "https://api.openai.com/v1";

// -------------------------
// Helper: Sanitize word input
// -------------------------
function sanitizeWord(text) {
  if (!text) return "";
  const cleaned = text.replace(/[^A-Za-z\- ]+/g, "").trim();
  const token = cleaned.split(" ")[0] || "";
  return token.toLowerCase();
}

// -------------------------
// Helper: Calculate cosine similarity
// -------------------------
function cosineSimilarity(vecA, vecB) {
  const dotProduct = vecA.reduce((sum, a, i) => sum + a * vecB[i], 0);
  const magnitudeA = Math.sqrt(vecA.reduce((sum, a) => sum + a * a, 0));
  const magnitudeB = Math.sqrt(vecB.reduce((sum, b) => sum + b * b, 0));
  return dotProduct / (magnitudeA * magnitudeB);
}

// -------------------------
// Helper: Scale similarity for better UX
// -------------------------
function scaleSimilarity(rawScore) {
  if (rawScore < 0.3) {
    return rawScore * 0.9;
  } else if (rawScore >= 0.3 && rawScore < 0.4) {
    return rawScore + 0.05;
  } else if (rawScore >= 0.4 && rawScore < 0.5) {
    const boost = 0.1 + (rawScore - 0.4) * 0.5;
    return rawScore + boost;
  } else if (rawScore >= 0.5 && rawScore < 0.6) {
    const boost = 0.15 + (rawScore - 0.5) * 0.5;
    return rawScore + boost;
  } else if (rawScore >= 0.6 && rawScore < 0.7) {
    const boost = 0.2 + (rawScore - 0.6) * 1.0;
    return rawScore + boost;
  } else {
    const boost = 0.3 + (rawScore - 0.7) * 0.5;
    return Math.min(rawScore + boost, 1.0);
  }
}

// -------------------------
// TIER 1: Word Bank - 50+ words per category (PRIMARY)
// -------------------------
const WORD_BANK = {
  music: [
    "guitar",
    "piano",
    "drums",
    "violin",
    "flute",
    "trumpet",
    "bass",
    "melody",
    "rhythm",
    "jazz",
    "blues",
    "rock",
    "concert",
    "band",
    "chorus",
    "tempo",
    "acoustic",
    "electric",
    "harmony",
    "tune",
    "song",
    "album",
    "vinyl",
    "stereo",
    "note",
    "beat",
    "chord",
    "scale",
    "octave",
    "voice",
    "vocal",
    "lyric",
    "artist",
    "singer",
    "composer",
    "bassoon",
    "ballads",
    "symphony",
    "opera",
    "disco",
    "folk",
    "country",
    "reggae",
    "funk",
    "grunge",
    "indie",
    "studio",
    "stage",
    "festival",
    "encore",
  ],
  sports: [
    "soccer",
    "football",
    "sweeper",
    "baseball",
    "tennis",
    "golf",
    "hockey",
    "rugby",
    "cricket",
    "boxing",
    "tackling",
    "swimming",
    "diving",
    "skating",
    "skiing",
    "surfing",
    "cycling",
    "running",
    "marathon",
    "sprint",
    "relay",
    "jump",
    "throw",
    "catch",
    "kick",
    "pass",
    "goal",
    "score",
    "point",
    "win",
    "team",
    "coach",
    "player",
    "athlete",
    "champion",
    "trophy",
    "medal",
    "victory",
    "court",
    "field",
    "arena",
    "stadium",
    "track",
    "pool",
    "rink",
    "gym",
    "fitness",
    "training",
    "workout",
    "exercise",
  ],
  food: [
    "pizza",
    "pasta",
    "burger",
    "salad",
    "taco",
    "sushi",
    "rice",
    "bread",
    "cheese",
    "chicken",
    "steak",
    "fish",
    "shrimp",
    "lobster",
    "crab",
    "bacon",
    "eggs",
    "toast",
    "waffle",
    "pancake",
    "cereal",
    "oatmeal",
    "yogurt",
    "fruit",
    "apple",
    "banana",
    "orange",
    "grape",
    "berry",
    "melon",
    "peach",
    "mango",
    "carrot",
    "broccoli",
    "pepper",
    "onion",
    "garlic",
    "potato",
    "tomato",
    "soup",
    "stew",
    "curry",
    "noodle",
    "dumpling",
    "sandwich",
    "wrap",
    "bagel",
    "cookie",
    "cake",
    "pie",
    "brownie",
  ],
  animals: [
    "elephant",
    "dolphin",
    "tiger",
    "lion",
    "bear",
    "wolf",
    "fox",
    "deer",
    "rabbit",
    "squirrel",
    "mouse",
    "rat",
    "hamster",
    "guinea",
    "giraffe",
    "zebra",
    "monkey",
    "gorilla",
    "panda",
    "koala",
    "kangaroo",
    "penguin",
    "eagle",
    "hawk",
    "owl",
    "parrot",
    "crow",
    "swan",
    "duck",
    "goose",
    "shark",
    "whale",
    "seal",
    "otter",
    "turtle",
    "frog",
    "snake",
    "lizard",
    "spider",
    "firefly",
    "bee",
    "ant",
    "ladybug",
    "cricket",
    "beetle",
    "puppy",
    "kitten",
    "horse",
    "cow",
    "sheep",
  ],
  nature: [
    "forest",
    "jungle",
    "desert",
    "mountain",
    "valley",
    "canyon",
    "cliff",
    "ocean",
    "sea",
    "lake",
    "river",
    "stream",
    "creek",
    "pond",
    "waterfall",
    "beach",
    "shore",
    "coast",
    "island",
    "volcano",
    "glacier",
    "iceberg",
    "tree",
    "pine",
    "oak",
    "maple",
    "birch",
    "willow",
    "palm",
    "bamboo",
    "flower",
    "rose",
    "lily",
    "daisy",
    "tulip",
    "orchid",
    "blossom",
    "grass",
    "moss",
    "fern",
    "vine",
    "bush",
    "shrub",
    "garden",
    "meadow",
    "sky",
    "cloud",
    "rain",
    "snow",
    "wind",
    "storm",
  ],
  technology: [
    "computer",
    "laptop",
    "tablet",
    "phone",
    "android",
    "screen",
    "keyboard",
    "mouse",
    "monitor",
    "printer",
    "scanner",
    "camera",
    "video",
    "audio",
    "software",
    "hardware",
    "program",
    "app",
    "code",
    "data",
    "file",
    "folder",
    "internet",
    "website",
    "email",
    "browser",
    "search",
    "cloud",
    "server",
    "network",
    "wifi",
    "wireless",
    "cable",
    "charger",
    "battery",
    "power",
    "digital",
    "virtual",
    "online",
    "stream",
    "download",
    "upload",
    "backup",
    "robot",
    "drone",
    "gadget",
    "device",
    "tech",
    "cyber",
    "gaming",
  ],
};

// -------------------------
// TIER 3: Expanded Fallback Words - 30+ per category (EMERGENCY)
// -------------------------
const FALLBACK_WORDS = {
  music: [
    "guitar",
    "piano",
    "drums",
    "violin",
    "melody",
    "rhythm",
    "jazz",
    "blues",
    "rock",
    "concert",
    "band",
    "chorus",
    "tempo",
    "tune",
    "song",
    "album",
    "vinyl",
    "stereo",
    "note",
    "beat",
    "chord",
    "harmony",
    "acoustic",
    "electric",
    "voice",
    "vocal",
    "lyric",
    "artist",
    "singer",
    "festival",
    "encore",
  ],
  sports: [
    "soccer",
    "tennis",
    "ballgame",
    "swimming",
    "football",
    "baseball",
    "golf",
    "hockey",
    "rugby",
    "boxing",
    "marathon",
    "sprint",
    "relay",
    "jump",
    "throw",
    "goal",
    "score",
    "team",
    "coach",
    "player",
    "athlete",
    "champion",
    "trophy",
    "medal",
    "victory",
    "stadium",
    "arena",
    "gym",
    "fitness",
    "workout",
  ],
  food: [
    "pizza",
    "pasta",
    "burger",
    "salad",
    "taco",
    "sushi",
    "rice",
    "bread",
    "cheese",
    "chicken",
    "steak",
    "fish",
    "bacon",
    "eggs",
    "toast",
    "waffle",
    "cereal",
    "yogurt",
    "apple",
    "banana",
    "orange",
    "grape",
    "berry",
    "mango",
    "carrot",
    "pepper",
    "soup",
    "curry",
    "noodle",
    "sandwich",
    "cookie",
    "cake",
  ],
  animals: [
    "elephant",
    "dolphin",
    "tiger",
    "penguin",
    "lion",
    "bear",
    "wolf",
    "fox",
    "rabbit",
    "giraffe",
    "zebra",
    "monkey",
    "panda",
    "koala",
    "kangaroo",
    "eagle",
    "hawk",
    "owl",
    "parrot",
    "shark",
    "whale",
    "seal",
    "turtle",
    "frog",
    "snake",
    "spider",
    "firefly",
    "bee",
    "puppy",
    "kitten",
    "horse",
  ],
  nature: [
    "forest",
    "ocean",
    "mountain",
    "river",
    "jungle",
    "desert",
    "valley",
    "canyon",
    "lake",
    "stream",
    "geyser",
    "beach",
    "island",
    "volcano",
    "glacier",
    "tree",
    "flower",
    "rose",
    "garden",
    "meadow",
    "sky",
    "cloud",
    "rain",
    "snow",
    "wind",
    "storm",
    "sunset",
    "sunrise",
    "moon",
    "star",
    "rainbow",
  ],
  technology: [
    "computer",
    "robot",
    "internet",
    "software",
    "laptop",
    "tablet",
    "phone",
    "screen",
    "keyboard",
    "mouse",
    "printer",
    "camera",
    "app",
    "code",
    "data",
    "website",
    "email",
    "browser",
    "cloud",
    "server",
    "network",
    "wifi",
    "digital",
    "virtual",
    "online",
    "download",
    "backup",
    "drone",
    "gadget",
    "gaming",
  ],
};

// -------------------------
// Function: Generate Word (3-TIER SYSTEM)
// -------------------------
export async function generateWord(category) {
  const categoryLower = category.toLowerCase();

  // ==================================================
  // TIER 1: Try Word Bank first (PRIMARY - 99% of time)
  // ==================================================
  const wordBankWords = WORD_BANK[categoryLower];
  if (wordBankWords && wordBankWords.length > 0) {
    const randomIndex = Math.floor(Math.random() * wordBankWords.length);
    const randomWord = wordBankWords[randomIndex];

    console.log(
      `✅ TIER 1 - Word Bank: "${randomWord}" (${wordBankWords.length} options available)`
    );

    return {
      category,
      word: randomWord,
      source: "wordbank-tier1",
    };
  }

  // ==================================================
  // TIER 2: Try OpenAI if Word Bank doesn't have category (BACKUP)
  // ==================================================
  if (OPENAI_API_KEY) {
    console.log("🤖 TIER 2 - Word Bank empty, trying OpenAI API...");

    const prompt = `Provide exactly one single English word related to the category '${category}'.
- The word should be 4 to 10 letters long if possible.
- Return ONLY the word, without explanation, punctuation, or quotes.`;

    try {
      const response = await fetch(`${OPENAI_API_URL}/chat/completions`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${OPENAI_API_KEY}`,
        },
        body: JSON.stringify({
          model: "gpt-4o-mini",
          messages: [{ role: "user", content: prompt }],
          temperature: 1.2,
          max_tokens: 15,
          presence_penalty: 0.6,
          frequency_penalty: 0.6,
        }),
      });

      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.error?.message || "OpenAI API error");
      }

      const word = sanitizeWord(data.choices[0].message.content);

      if (word) {
        console.log(`✅ TIER 2 - OpenAI generated: "${word}"`);
        return {
          category,
          word,
          source: "openai-tier2",
        };
      }
    } catch (error) {
      console.error("⚠️ TIER 2 - OpenAI failed:", error.message);
    }
  }

  // ==================================================
  // TIER 3: Fallback to expanded word list (EMERGENCY)
  // ==================================================
  console.log("📦 TIER 3 - Using fallback word list (emergency)");
  const fallbackWordList = FALLBACK_WORDS[categoryLower] || [
    "word",
    "game",
    "play",
    "star",
    "moon",
  ];
  const randomIndex = Math.floor(Math.random() * fallbackWordList.length);
  const randomWord = fallbackWordList[randomIndex];

  console.log(
    `✅ TIER 3 - Fallback: "${randomWord}" (${fallbackWordList.length} options available)`
  );

  return {
    category,
    word: randomWord,
    source: "fallback-tier3",
  };
}

// -------------------------
// Function: Get Embedding
// -------------------------
async function getEmbedding(text) {
  if (!OPENAI_API_KEY) {
    throw new Error("OPENAI_API_KEY not set in environment");
  }

  const response = await fetch(`${OPENAI_API_URL}/embeddings`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${OPENAI_API_KEY}`,
    },
    body: JSON.stringify({
      model: "text-embedding-3-large",
      input: text,
    }),
  });

  const data = await response.json();

  if (!response.ok) {
    throw new Error(data.error?.message || "OpenAI embeddings API error");
  }

  return data.data[0].embedding;
}

// -------------------------
// Function: Calculate Similarity
// -------------------------
export async function calculateSimilarity(word1, word2) {
  const w1 = sanitizeWord(word1);
  const w2 = sanitizeWord(word2);

  if (!w1 || !w2) {
    throw new Error("Both word1 and word2 are required");
  }

  try {
    const [embedding1, embedding2] = await Promise.all([
      getEmbedding(w1),
      getEmbedding(w2),
    ]);

    const rawScore = cosineSimilarity(embedding1, embedding2);
    const scaledScore = scaleSimilarity(rawScore);

    return {
      word1: w1,
      word2: w2,
      similarity: Math.round(scaledScore * 1000) / 1000,
      raw_similarity: Math.round(rawScore * 1000) / 1000,
      source: "openai-text-embedding-3-large",
    };
  } catch (error) {
    console.error("Similarity calculation error:", error);
    throw error;
  }
}
